File: src/neml2/models/solid_mechanics/SwellingAndPhaseChangeDeformationJacobian.cxx
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "neml2/models/solid_mechanics/SwellingAndPhaseChangeDeformationJacobian.h"
#include "neml2/tensors/Scalar.h"

namespace neml2
{
register_NEML2_object(SwellingAndPhaseChangeDeformationJacobian);

OptionSet
SwellingAndPhaseChangeDeformationJacobian::expected_options()
{
  OptionSet options = Model::expected_options();
  options.doc() =
      "Define the linear isotropic phase change deformation Jacobian for a freezing liquid or a "
      "melting solid, i.e. \\f$ J = \\left( 1 + \\alpha c \\phi^f + (1-c) \\phi^f \\Delta \\Omega "
      "\\right) \\f$, where \\f$ \\alpha \\f$ is the coefficient of swelling, \\f$ \\Delta \\Omega "
      "\\f$ is relative difference of the reference volume between the two phases, \\f$ \\phi^f "
      "\\f$ "
      "is the fluid fraction associated with swelling, and \\f$ c \\f$ is the phase fraction.";

  options.set_input("fluid_fraction") = VariableName(STATE, "phi_f");
  options.set("fluid_fraction").doc() = "Volume fraction of the fluid phase.";

  options.set_parameter<TensorName<Scalar>>("phase_fraction");
  options.set("phase_fraction").doc() =
      "Phase fraction during the phase change. 0 means all solid, 1 means all liquid.";

  options.set_parameter<TensorName<Scalar>>("swelling_coefficient");
  options.set("swelling_coefficient").doc() = "Coefficient of phase expansion";

  options.set_parameter<TensorName<Scalar>>("reference_volume_difference");
  options.set("reference_volume_difference").doc() =
      "Relative difference between the reference volumes of the two phases.";

  options.set<VariableName>("jacobian") = VariableName(STATE, "J");
  options.set("jacobian").doc() = "Phase change deformation Jacobian";

  return options;
}

SwellingAndPhaseChangeDeformationJacobian::SwellingAndPhaseChangeDeformationJacobian(
    const OptionSet & options)
  : Model(options),
    _vf(declare_input_variable<Scalar>("fluid_fraction")),
    _c(declare_parameter<Scalar>("c", "phase_fraction", /*allow_nonlinear=*/true)),
    _alpha(declare_parameter<Scalar>("alpha", "swelling_coefficient")),
    _dOmega(declare_parameter<Scalar>("dOmega", "reference_volume_difference")),
    _J(declare_output_variable<Scalar>("jacobian"))
{
}

void
SwellingAndPhaseChangeDeformationJacobian::set_value(bool out, bool dout_din, bool /*d2out_din2*/)
{
  if (out)
    _J = (1.0 + _alpha * _c * _vf + (1 - _c) * _vf * _dOmega);

  if (dout_din)
  {
    if (_vf.is_dependent())
      _J.d(_vf) = (_alpha * _c + (1 - _c) * _dOmega);

    if (const auto * const c = nl_param("c"))
      _J.d(*c) = (_alpha * _vf - _vf * _dOmega);
  }
}
} // namespace neml2

File: neml2/models/solid_mechanics/kinematics/SwellingAndPhaseChangeDeformationJacobian.py
# Copyright 2024, UChicago Argonne, LLC
# All Rights Reserved
# Software Name: NEML2 -- the New Engineering material Model Library, version 2
# By: Argonne National Laboratory
# OPEN SOURCE LICENSE (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

"""Python-native mirror of C++ ``SwellingAndPhaseChangeDeformationJacobian`` (kinematics)."""

from __future__ import annotations

from ....chain_rule import ChainRuleDict
from ....factory import register_native
from ....model import Model
from ....schema import HitSchema, input, output, parameter
from ....types import Scalar


@register_native("SwellingAndPhaseChangeDeformationJacobian")
class SwellingAndPhaseChangeDeformationJacobian(Model):
    r"""Define the linear isotropic phase change deformation Jacobian for a freezing liquid or a
    melting solid, i.e. $J = \left( 1 + \alpha c \phi^f + (1-c) \phi^f \Delta \Omega \right)$, where
    $\alpha$ is the coefficient of swelling, $\Delta \Omega$ is relative
    difference of the reference volume between the two phases, $\phi^f$ is the fluid
    fraction associated with swelling, and $c$ is the phase fraction.
    """

    # Forward: ``J = 1 + alpha * c * vf + (1 - c) * vf * dOmega``. Linear in
    # the fluid_fraction input (action = (alpha*c + (1-c)*dOmega) * V) and,
    # when the phase_fraction parameter is promoted to a nonlinear input, also
    # linear in c (action = (alpha*vf - vf*dOmega) * V). Pure typed wrapper
    # algebra throughout: Scalar + - * handle sub-batch alignment, so
    # no torch.<op>(.data) calls appear in forward or chain-rule action bodies.

    hit = HitSchema(
        input("fluid_fraction", Scalar, "Volume fraction of the fluid phase."),
        output("jacobian", Scalar, "Phase change deformation Jacobian"),
        parameter(
            "phase_fraction",
            Scalar,
            "Phase fraction during the phase change. 0 means all solid, 1 means all liquid.",
            attr="c",
            allow_nonlinear=True,
        ),
        parameter(
            "swelling_coefficient",
            Scalar,
            "Coefficient of phase expansion",
            attr="alpha",
        ),
        parameter(
            "reference_volume_difference",
            Scalar,
            "Relative difference between the reference volumes of the two phases.",
            attr="dOmega",
        ),
    )

    # ``from_hit`` auto-declares ``phase_fraction`` (stored as ``c``),
    # ``swelling_coefficient`` (stored as ``alpha``), and
    # ``reference_volume_difference`` (stored as ``dOmega``). Annotate so
    # pyright sees the typed wrappers that ``Model.__getattr__`` returns
    # rather than the inherited ``nn.Module`` hint.
    c: Scalar
    alpha: Scalar
    dOmega: Scalar

    def forward(  # type: ignore[override]
        self,
        fluid_fraction: Scalar,
        *nl_params,
        v: ChainRuleDict | None = None,
    ):
        vf = fluid_fraction
        c = self._get_param("c", nl_params, Scalar)
        alpha = self.alpha
        dOmega = self.dOmega
        J = 1.0 + alpha * c * vf + (1.0 - c) * vf * dOmega
        if v is None:
            return J

        # Differential pushforward: linear in vf (and, when promoted,
        # linear in c). Each action is the same wrapper map applied to the
        # incoming Scalar tangent V scaled by the held coefficient. No
        # Jacobian materialised. Mirrors the C++
        # ``_J.d(_vf) = (alpha*c + (1-c)*dOmega)`` and
        # ``_J.d(c) = (alpha*vf - vf*dOmega)`` derivatives.
        dJ_dvf = alpha * c + (1.0 - c) * dOmega

        def fluid_fraction_action(V: Scalar, coef=dJ_dvf) -> Scalar:
            return coef * V

        actions: dict = {"fluid_fraction": fluid_fraction_action}

        if "c" in self._nl_params:
            dJ_dc = alpha * vf - vf * dOmega
            actions[self._nl_params["c"].input_name] = lambda V, coef=dJ_dc: coef * V

        return J, self.apply_chain_rule(v, "jacobian", actions, output=J)


__all__ = ["SwellingAndPhaseChangeDeformationJacobian"]
